### ISSUE TITLE
Fyne does not compile on FreeBSD

### ISSUE THREAD

**Issue by ruesch** [fyne-io/fyne#2130]
Hello,

I am having problems getting fyne to work on FreeBSD. I do not know if this is a bug or if I have forgotten something and ask you here for help.

I have set up a fresh FreeBSD and installed the dependencies as described on https://developer.fyne.io/started/ with `sudo pkg install go gcc xorg glfw pkgconf`.

After that I proceeded as follows, where the installation aborted with an error message:
```
$ go get fyne.io/fyne/v2
go: downloading fyne.io/fyne/v2 v2.0.2
go: downloading fyne.io/fyne v1.4.3

$ go get fyne.io/fyne/v2/cmd/fyne_demo
go: downloading github.com/srwiley/oksvg v0.0.0-20200311192757-870daf9aa564
go: downloading github.com/srwiley/rasterx v0.0.0-20200120212402-85cb7272f5e9
go: downloading golang.org/x/image v0.0.0-20200430140353-33d19683fad8
go: downloading github.com/goki/freetype v0.0.0-20181231101311-fa8a33aabaff
go: downloading github.com/fsnotify/fsnotify v1.4.9
go: downloading github.com/fyne-io/mobile v0.1.3-0.20210318200029-09e9c4e13a8f
go: downloading github.com/godbus/dbus/v5 v5.0.3
go: downloading golang.org/x/sys v0.0.0-20200720211630-cb9d2d5c5666
go: downloading github.com/fredbi/uri v0.0.0-20181227131451-3dcfdacbaaf3
go: downloading github.com/stretchr/testify v1.5.1
go: downloading github.com/go-gl/glfw/v3.3/glfw v0.0.0-20210311203641-62640a716d48
go: downloading github.com/go-gl/gl v0.0.0-20190320180904-bf2b1f2f34d7
go: downloading gopkg.in/yaml.v2 v2.2.8
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading golang.org/x/net v0.0.0-20200324143707-d3edc9973b7e
go: downloading golang.org/x/text v0.3.2
# github.com/godbus/dbus/v5
In file included from go/pkg/mod/github.com/godbus/dbus/v5@v5.0.3/transport_unixcred_freebsd.go:13:
In file included from /usr/include/sys/ucred.h:39:
/usr/include/sys/_lock.h:38:2: error: unknown type name 'u_int'
        u_int   lo_flags;
        ^
/usr/include/sys/_lock.h:39:2: error: unknown type name 'u_int'
        u_int   lo_data;                /* General class specific data. */
        ^
In file included from go/pkg/mod/github.com/godbus/dbus/v5@v5.0.3/transport_unixcred_freebsd.go:13:
In file included from /usr/include/sys/ucred.h:40:
/usr/include/sys/_mutex.h:49:11: error: unknown type name 'uintptr_t'
        volatile uintptr_t      mtx_lock;       /* Owner and flags. */
                 ^
/usr/include/sys/_mutex.h:63:11: error: unknown type name 'uintptr_t'
        volatile uintptr_t      mtx_lock;       /* Owner and flags. */
                 ^
4 errors generated.
# github.com/go-gl/glfw/v3.3/glfw
In file included from go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20210311203641-62640a716d48/native_linbsd.go:10:
In file included from go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20210311203641-62640a716d48/glfw/include/GLFW/glfw3native.h:114:
In file included from /usr/local/include/GL/glx.h:32:
/usr/local/include/GL/gl.h:56:11: warning: 'GLAPIENTRY' macro redefined [-Wmacro-redefined]
go/pkg/mod/github.com/go-gl/glfw/v3.3/glfw@v0.0.0-20210311203641-62640a716d48/glfw/include/GLFW/glfw3.h:5894:10: note: previous definition is here
```

**Comment by andydotxyz** [comment]
This should work, we have a live CD using FreeBSD, so I know that it compiled a few weeks back.
@Jacalz can you comment on this? I think you're more frequently on FreeBSD than I am?

**Comment by andydotxyz** [comment]
Just tested and it's working great. Please can you provide the information about system versions etc (the bug report template should have asked you these questions).

**Comment by Jacalz** [comment]
Hmm, I have not had any issues getting it to compile on FreeBSD, but most of my testing has been done on GhostBSD. The logs seem to indicate that you don’t have dbus installed (the glfw stuff is just warnings, it works without it). What desktop environment are you using?

**Comment by ruesch** [comment]
Hello,

thanks for your answers and sorry for the poor bug report. It is the first time that I post a bug-report. Here are the missing infos:

OS: FreeBSD 13.0 RC4 - amd64
CPU: AMD Ryzen 9 3900X
GPU: NVIDIA GeForce RTX 2080 SUPER (nvidia-driver-460.67)
go version go1.16.2 freebsd/amd64

On the same PC it runs under Linux and under Windows 10.

The same problem exists on another PC:

OS: FreeBSD 13.0 RC2 - amd64
CPU: AMD Ryzen 5 3400GE
GPU: AMD Radeon Vega 11 graphics (drm-kmod-g20190710_1 with amdgpu kernel module loaded)
go version go1.16.2 freebsd/amd64

dbus is running:
messagebus 87104    0.0  0.0    14044   4052  -  Is   14:18       0:00.17 /usr/local/bin/dbus-daemon --system

I will test the whole thing under GhostBSD. Maybe it is a problem with FreeBSD 13, which will be released the next days, or a problem with Ryzen processors?

**Comment by Jacalz** [comment]
Thanks for sharing the extra data.

I highly doubt that it would be an issue with Ryzen processors. The errors looks more like something changed with the C-library or a C compiler. That leads me to think, like you said, that it could be specific to FreeBSD 13, as all my testing has been on 12.1 and 12.2. I did find https://github.com/znc/znc/issues/1313 that seems to have had a similar issue a while back and an article with a possible solution https://www.oleherman.com/dev/unknown_type_name/.

**Comment by ruesch** [comment]
Here is a small status update. I tested FreeBSD 12.2 today and fyne compiles here without any problems. It really looks like there are changes in FreeBSD 13.0 that affect the compilation of fyne. Unfortunately, due to my PC's hardware support, I can't go back to FreeBSD 12.2. It's not an urgent problem for me, since I'm new to fyne and can also switch to Linux.

**Comment by andydotxyz** [comment]
Is this an expected change in 13.0, or might it be a bug? If you can find any documentation about the change then we can fix it. Unfortunately we’re not really able to test on pre-releases due to the high number of OSes we already have to support.

**Comment by ruesch** [comment]
It looks like the problem is not with fyne, but with godbus / dbus.

```
$ go get github.com/godbus/dbus
go: downloading github.com/godbus/dbus v4.1.0+incompatible
# github.com/godbus/dbus
In file included from go/pkg/mod/github.com/godbus/dbus@v4.1.0+incompatible/transport_unixcred_freebsd.go:13:
In file included from /usr/include/sys/ucred.h:39:
/usr/include/sys/_lock.h:38:2: error: unknown type name 'u_int'
        u_int   lo_flags;
        ^
/usr/include/sys/_lock.h:39:2: error: unknown type name 'u_int'
        u_int   lo_data;                /* General class specific data. */
        ^
In file included from go/pkg/mod/github.com/godbus/dbus@v4.1.0+incompatible/transport_unixcred_freebsd.go:13:
In file included from /usr/include/sys/ucred.h:40:
/usr/include/sys/_mutex.h:49:11: error: unknown type name 'uintptr_t'
        volatile uintptr_t      mtx_lock;       /* Owner and flags. */
                 ^
/usr/include/sys/_mutex.h:63:11: error: unknown type name 'uintptr_t'
        volatile uintptr_t      mtx_lock;       /* Owner and flags. */
                 ^
4 errors generated.
```
https://github.com/godbus/dbus/issues/228
https://github.com/godbus/dbus/issues/225

I am trying to figure out what has changed since 12.2. As soon as I know more, I'll let you know.

Thank you for your great and fast support. :-)

**Comment by Jacalz** [comment]
Sounds more like it’s actually an issue with the libraries imported by godbus, not specifically an issue within their code. It might be that they need add the definition mentioned in https://www.oleherman.com/dev/unknown_type_name/.

**Comment by ruesch** [comment]
Hello,

i found the problem. fyne installs the package dbus in version 5.0.3. In this version the `#include <sys/types.h>` is missing in `transport_unixcred_freebsd.go`. If you look on [,](https://github.com/godbus/dbus/blob/master/transport_unixcred_freebsd.go) this definition is present in the newer version. After I added this by hand, fyne compiles again.

**Comment by Jacalz** [comment]
Nice work @ruesch! We should probably try and update godbus to 5.0.4 then 🙂

**Comment by andydotxyz** [comment]
Should be fixed on `develop` and `release/v2.0.x` branches, ready for whichever release is next
